src/Data/TypedByteArray.hs

```diff
@@ -36,13 +51,27 @@ unsafeIndex (TypedByteArray arr) = indexByteArray arr
 
 {-# INLINE generate #-}
 -- | Construct a 'TypedByteArray' of the given length by applying the function to each index in @[0..n-1]@.
-generate :: forall a. Prim a => Int -> (Int -> a) -> TypedByteArray a
+generate :: Prim a => Int -> (Int -> a) -> TypedByteArray a
 generate !n f = runST $ do
     -- Allocate enough space for n elements of type a
-    arr <- newByteArray $ sizeOf (undefined :: a) * n
-    intLoop 0 n $ \i -> i `seq` writeByteArray arr i $ f i
+    arr <- newTypedByteArray n
+    intLoop 0 n $ \i -> i `seq` writeTypedByteArray arr i $ f i
```

Comment:
The parameter is made strict explicitly because this piece of code is very annoying to view in Core because of inlining and the `ST` stuff around it. TIL that GHC doesn't like `\!i -> ...` but `\(!i) -> ...` is okay.